foyer/xml_writer.py

```diff
@@ -0,0 +1,281 @@
+    for small molecules to be used to test the implementation of a Foyer
+    force field XML.
+
+    Additional features to be added include:
+    - Documentation (and better standardization) of parameter rounding.
+    - Grouping of duplicate parameter definitions by atom class (the current
+      implementation considers parameters to be unique if atom `type`
+      definitions differ)
+    - Sort parameters by atom ids when `unique=False` (currently no sorting
+      is performed, this is minor but would make files slightly more readable)
+
+    Parameters
+    ----------
+    filename : str
+        Name of the Foyer XML file to be written
+    forcefield : foyer.Forcefield, optional, default=None
+        Foyer Forcefield used to parameterize the ParmEd Structure. This
+        is used to obtain additional information that is not available
+        from the Structure itself, e.g. atomtype SMARTS definitions.
+    unique : boolean, optional, default=True
+        Write only unique elements. If False, elements are written for each
+        atom, bond, etc. in the system. `unique=False` is primarily used
+        for storing the topology of "test" molecules for a Foyer forcefield.
+
+    """
+    # Assume if a Structure has a bond and bond type that the Structure is
+    # parameterized. ParmEd uses the same logic to denote parameterization.
+    if not (len(self.bonds) > 0 and self.bonds[0].type is not None):
+        raise Exception("Cannot write Foyer XML from an unparametrized "
+                        "Structure.")
+
+    root = ET.Element('ForceField')
+    _write_atoms(self, root, self.atoms, forcefield, unique)
+    if len(self.bonds) > 0 and self.bonds[0].type is not None:
+        _write_bonds(root, self.bonds, unique)
+    if len(self.angles) > 0 and self.angles[0].type is not None:
+        _write_angles(root, self.angles, unique)
+    if len(self.dihedrals) > 0 and self.dihedrals[0].type is not None:
+        _write_periodic_torsions(root, self.dihedrals, unique)
+    if len(self.rb_torsions) > 0 and self.rb_torsions[0].type is not None:
+        _write_rb_torsions(root, self.rb_torsions, unique)
+
+    _remove_duplicate_elements(root, unique)
+
+    ET.ElementTree(root).write(filename, pretty_print=True)
+
+
+def _write_atoms(self, root, atoms, forcefield, unique):
+    atomtypes = ET.SubElement(root, 'AtomTypes')
+    nonbonded = ET.SubElement(root, 'NonbondedForce')
+    nonbonded.set('coulomb14scale', str(_infer_coulomb14scale(self)))
+    nonbonded.set('lj14scale', str(_infer_lj14scale(self)))
+    atomtype_attrs = collections.OrderedDict([
+        ('name', 'name'),
+        ('class', 'forcefield.atomTypeClasses[name]'),
+        ('element', 'forcefield.atomTypeElements[name]'),
+        ('mass', 'atom.mass'),
+        ('def', 'forcefield.atomTypeDefinitions[name]'),
+        ('desc', 'forcefield.atomTypeDesc[name]'),
+        ('doi', 'forcefield.atomTypeRefs[name]')
```

Comment:
Doing the iteration through all these XML attributes, the resultant XML looks like
`<Type name="opls_135" class="CT" element="C" mass="12.01078" def="[C;X4](C)(H)(    H)H" desc="" doi="{&quot;{'10.1021/ja9621760'}&quot;}"/>`

When it should be 
` <Type name="opls_135" class="CT" element="C" mass="12.01078" def="[C;X4](C)(H)(    H)H" desc="" doi="{'10.1021/ja9621760'}"/>`

There's some sort of weird formatting for the doi, I wonder if this can be done by stripping out the `{&quot;` from the doi attribute